File: algorithms/python/grumpyBookstoreOwner/grumpyBookstoreOwner.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/grumpy-bookstore-owner/
# Author: Miao Zhang
# Date:   2021-04-07

class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], X: int) -> int:
        base = 0
        window = 0
        best_window = 0
        for i in range(len(grumpy)):
            if grumpy[i]:
                window += customers[i]
            else:
                base += customers[i]
            if i >= X and grumpy[i - X]:
                window -= customers[i - X]
            best_window = max(best_window, window)
        return base + best_window

File: algorithms/cpp/grumpyBookstoreOwner/grumpyBookstoreOwner.cpp
// Source: https://leetcode.com/problems/grumpy-bookstore-owner/
// Author: Miao Zhang
// Date:   2021-04-07

class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int X) {
        int base = 0;
        int window = 0;
        int best_window = 0;
        for (int i = 0; i < grumpy.size(); i++) {
            (grumpy[i] ? window : base) += customers[i];
            if (i >= X && grumpy[i - X])
                window -= customers[i - X];
            best_window = max(best_window, window);
        }
        return base + best_window;
    }
};
